### PR TITLE
Add depends_on to the input manifest and increment its schema to 1.1

### PR DESCRIPTION
**Pull request by zelinh**
### Description
Add depends_on to the input manifest and increment its schema to 1.1.

### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/4157

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/build_workflow/builders.py [hunk 1]
```diff
@@ -11,15 +11,15 @@
...
 from build_workflow.builder_from_source import BuilderFromSource
-from manifests.input_manifest import InputComponent, InputComponentFromDist, InputComponentFromSource
+from manifests.input_manifest import InputComponent
 
 
 class Builders(ABC):
     @classmethod
     def builder_from(self, component: InputComponent, target: BuildTarget) -> Builder:
-        if type(component) is InputComponentFromDist:
+        if hasattr(component, "dist"):
```
Comment (dblock) [inline src/build_workflow/builders.py:20]: This causes the need to do a `type: ignore` below, why is this needed?

Comment (zelinh) [reply src/build_workflow/builders.py:20]: @dblock Good point! Those below will give us `error: "InputComponent" has no attribute "repository"  [attr-defined]`.
I think the reason is that it's iterating through `manifest.components.select` and that method is set to return an iterator of `InputComponent` type even though we have the check to make sure it has `repository` attribute.

Comment (rishabh6788) [reply src/build_workflow/builders.py:20]: Little confused too, why it can't be handled with existing implementation?

Comment (zelinh) [reply src/build_workflow/builders.py:20]: I think the usages of `component` are different in these two scenarios. In this `builders.py` file, we don't iterate through components and don't access any of the attribute. 

In below example from `run_checkout.py`
```
        for component in manifest.components.select():
            logging.info(f"Checking out {component.name}")
            if hasattr(component, "repository"):
                with GitRepository(
                    component.repository,  # type: ignore[attr-defined]
                    component.ref,  # type: ignore[attr-defined]
                    os.path.join(work_dir.name, component.name),
                    component.working_directory,  # type: ignore[attr-defined]
                ) as repo:
                    logging.debug(f"Checked out {component.name} into {repo.dir}")
```
We are getting an iterator of `InputComponent` type. Even though we have a conditional statement 
`if hasattr(component, "repository")`, python type check seems not able to tell it's a `InputComponentFromSource` and when we try to access its attribute such as `component.repository`, it says `error: "InputComponent" has no attribute "repository"  [attr-defined]`.

The workflow can be run correctly with no error when I was testing on them and the python tests are passing as well.

Comment (rishabh6788) [reply src/build_workflow/builders.py:20]: As per my understanding you are only adding a new `depends_on` field in the input manifest schema. This change should not break the existing code flow, ideally the component object type would always be either `InputComponentFromSource` or `InputComponentFromDist`. 
What is the error you are getting with your changes, is it related to type incompatibility?
Please add the error you are facing with type check implementation, we can work from there to figure out what is going on. 
@dblock Do you think this is the right way forward?

Comment (dblock) [reply src/build_workflow/builders.py:20]: 100%, this change is unexpected

Comment (zelinh) [reply src/build_workflow/builders.py:20]: Yeah this was unexpected for me at first as well. When looking into the reason, this change is actually to support backward compatibility. 
Because when we incrementing the schema version to `1.1`, we keep the previous input_manifest class as `input_manifest_1_0` to support any previous 1.0 manifest. 
If in this case `builders` class, we are building `1.0` version input manifest, the condition statement would never fulfilled because the component object types are either `InputComponentFromSource_1_0` or `InputComponentFromDist_1_0` imported from `Input_Manifest_1_0`. 
These changes are making the conditional check more generic and sustainable for future schema version change for `input_manifest` as well. 
Let me know what you think about. Thanks! @rishabh6788 @dblock

File: src/manifests/input/input_manifest_1_0.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+        :param List[str] focus: Choose some components.
+        :param str platform: Only components targeting a given platform.
+        :return: Collection of components.
+        :raises ValueError: Invalid platform or component name specified.
+        """
+        if focus and len(focus) > 0:
+            invalid = [item for item in focus if item not in self]
+            if len(invalid) > 0:
+                raise ValueError(f"Unknown component{'s'[:len(invalid) != 1]}={','.join(invalid)}.")
```
Comment (rishabh6788) [inline src/manifests/input/input_manifest_1_0.py:191]: please add missing tests.

File: src/manifests/input/input_manifest_1_0.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+        self.checks = list(map(lambda entry: Check_1_0(entry), data.get("checks", [])))
+
+    @classmethod
+    def _from(self, data: dict) -> 'InputComponent_1_0':
+        if "repository" in data:
+            return InputComponentFromSource_1_0(data)
+        elif "dist" in data:
+            return InputComponentFromDist_1_0(data)
+        else:
+            raise ValueError(f"Invalid component data: {data}")
```
Comment (rishabh6788) [inline src/manifests/input/input_manifest_1_0.py:215]: please add missing tests.

File: src/manifests/input/input_manifest_1_0.py (hunk 2 above)
Comment (zelinh) [reply src/manifests/input/input_manifest_1_0.py:191]: These functions are basically the same as the main `input_manifest` class, just the previous `1.0` schema version. However I did add some tests for different schema versions in `tests/tests_manifests/test_input_manifest.py`. 
I don't see we have separate tests for old schema version `build_manifest` or `bundle_manifest`. Do we want to create duplicate tests for this class especially?

File: src/manifests/input/input_manifest_1_0.py (hunk 3 above)
Comment (zelinh) [reply src/manifests/input/input_manifest_1_0.py:215]: Replied above. ^^

File: src/build_workflow/builders.py (hunk 1 above)
Comment (rishabh6788) [reply src/build_workflow/builders.py:20]: Also do you think adding an optional field warrant a new minor version upgrade in the schema? 
What is the disadvantage of editing the existing schema and adding the optional `depends_on` field in the existing input manifest schema?
@zelinh @dblock

Comment (zelinh) [reply src/build_workflow/builders.py:20]: > @zelinh Can you try making it `type(component) in ['InputComponentFromSource_1_0', 'InputComponentFromSource']` and try? Same for `Dist`

I don't think it's sustainable if we do this. We would need to add it each time if we have a new schema version of the component type.

Comment (zelinh) [reply src/build_workflow/builders.py:20]: > Also do you think adding an optional field warrant a new minor version upgrade in the schema? What is the disadvantage of editing the existing schema and adding the optional `depends_on` field in the existing input manifest schema? @zelinh @dblock

Basically we are adding a new field in our input manifest so typically we would upgrade to a new schema version and keep the previous version in a sub-class. For example our `build_manifest` : https://github.com/opensearch-project/opensearch-build/tree/main/src/manifests/build.
If we are modifying the existing schema without incrementing a version, I'm worried that it may cause backward incompatibility in some unexpected cases on other workflows (e.g. build/ci workflow)

Comment (rishabh6788) [reply src/build_workflow/builders.py:20]: > > @zelinh Can you try making it `type(component) in ['InputComponentFromSource_1_0', 'InputComponentFromSource']` and try? Same for `Dist`
> 
> I don't think it's sustainable if we do this. We would need to add it each time if we have a new schema version of the component type.

I am okay to approve this if there is no other way to fix apart from type-ignore.
@dblock LMK your thoughts.

Comment (zelinh) [reply src/build_workflow/builders.py:20]: Tagging @gaiksaya for the concerns here.

Comment (dblock) [reply src/build_workflow/builders.py:20]: I think either all `InputComponentFromDist*` classes need a common parent class (`InputComponentFromDist`), or you can explicitly check for `InputComponentFromDist` and `InputComponentFromDist_1_0` here?

**Comment by dblock** [comment]
One thought. The `depends_on` feature is a net, optional, addition. Why did you decide to increment the schema version for it - can it be done as such on the existing schema?

File: src/build_workflow/builders.py (hunk 1 above)
Comment (zelinh) [reply src/build_workflow/builders.py:20]: @dblock I also thought about having a common parent class such as `InputComponentFromDist`. While the current parent class of `InputComponentFromDist/InputComponentFromDist_1_0` are `InputComponent` and `InputComponent_1_0` which inherit from generic `Component` so it's kinda the same situation for them. 
We could maybe create a new general class as the parent class but I think it may create more chaos. 

I think importing both `InputComponentFromDist` and `InputComponentFromDist_1_0` could work but in long term, we may have to add more and more class for type checking and it's not easy to maintain.

**Review by dblock — APPROVED** [review]
I tried to come up with a different approach, but didn't do better. This LGTM!
